### DISCUSSION TITLE
Can't seem to use template inheritance from Tera.

### DISCUSSION THREAD

**Discussion by danielyrovas** [rwf2/Rocket#1892] [Questions] [answered]
I have launched a rocket and I get the following error when I access the route:
```
parsing errors: Template 'navbar.html.tera' is inheriting from 'base', which doesn't exist or isn't loaded.
```

I have these templates in the default template directory:
<details>
<summary>base.html.tera</summary>

```html
<!DOCTYPE html>
<html lang="en">
<head>
    {% block head %}
    <meta charset="utf-8" />
    <link rel="stylesheet" href="style/style.css" />
    <title>{% block title %}{% endblock title %} - My Webpage</title>
    {% endblock head %}
</head>
<body>
    <nav>{% block navbar %}{% endblock navbar %}</nav>
    <main>{% block main %}{% endblock main %}</main>
    <footer>{% block footer %}{% endblock footer %}</footer>
</body>
</html>
```
</details>

<details>
<summary>navbar.html.tera</summary>

```html
{% extends "base" %}
{% block navbar %}
<nav class="menu">
    <a class="focus" href="index.html">Home</a>
</nav>
{% endblock navbar %}
```
</details>
my rust code is as follows
<details>
<summary>main.rs</summary>

```rust
#![cfg_attr(debug_assertions, allow(dead_code, unused_imports))]
#[cfg(test)] mod tests;

#[macro_use] extern crate rocket;
use std::{collections::HashMap, task::Context};

use rocket::{
    Request, Data, Response,
    fairing::{Fairing, Info, Kind},
    http::{Method, ContentType, Status},
    fs::{FileServer, relative}
};
use rocket_dyn_templates::{
    Template,
    tera::{Tera}
};

#[get("/")]
fn index() -> Template {
    let mut context = HashMap::new();
    context.insert( "base", "navbar");
    Template::render("base", &context)
}

#[launch]
fn rocket() -> _ {
    let tera = match Tera::new("templates/**/*") {
        Ok(t) => t,
        Err(e) => {
            println!("parsing errors: {}", e);
            ::std::process::exit(1);
        }
    };
    let names: Vec<_> = tera.get_template_names().collect();
    println!("{:?}", names);
    rocket::build().mount("/", routes![
        index,
        ])
        .mount("/", FileServer::from(relative!("static")))
        .attach(Template::fairing())
}
```
</details>

I have looked at Tera's documentation, and the Rocket Template section, but I can't find anything that indicates what I have done wrong. 
I can setup a rocket to server templates when I don't use inheritance.
I have also tried  changing the ``` {% extends "base" %}``` to ```{% extends "base.html" %}``` but that produces the same result. Changing to  ```{% extends "base.html.tera" %}``` produces a panic instead:
```cargo
Error: Failed to initialize Tera templating.
   >> Template 'navbar' is inheriting from 'base.html.tera', which doesn't exist or isn't loaded.
   >> Template initialization failed. Aborting launch.
Error: Rocket failed to launch due to failing fairings:
   >> Templating
```

**Comment by danielyrovas** [comment]
Also this is my Cargo.toml
```TOML
[package]
name = "hello_rocket"
version = "0.1.1"
edition = "2018"

[dependencies]
rocket = { version = "0.5.0-rc.1", features = ["json"] }
rocket_dyn_templates = { version = "0.1.0-rc.1", features = ["tera"] }
diesel = { version = "1.4.4", features = ["postgres"] }
dotenv = "0.15.0"
```
Thanks if you have an answer :)

**Comment by jebrosen** [comment, marked as answer]
> `parsing errors: Template 'navbar.html.tera' is inheriting from 'base', which doesn't exist or isn't loaded.`

This error is showing because `Tera::new()` and `rocket_dyn_templates` use different rules for file extensions. `Tera::new()` strips one file extension (or none, it seems); [`rocket_dyn_templates` strips two](https://api.rocket.rs/master/rocket_dyn_templates/index.html#naming).

(Aside from that, your templates don't seem quite right. `extends` is used for template inheritance, but this navbar looks like a use case for `include` instead.)

**Reply by danielyrovas** [reply]
yeah thanks
I think I'm going to switch to handlebars, it seems a bit more featured and the rust-lang website is a good place to start for my own project.
Can I use ```handlebars::new()``` and then set things like strict mode and custom helpers? Or would I need to register a custom template engine?

**Reply by jebrosen** [reply]
> Can I use handlebars::new() and then set things like strict mode and custom helpers? Or would I need to register a custom template engine?

Sure! You could use `Handlebars::new()` and store/access it through [managed state](https://rocket.rs/v0.5-rc/guide/state/), or you could use `rocket_dyn_templates`'s [`Template::custom()`](https://api.rocket.rs/v0.5-rc/rocket_dyn_templates/struct.Template.html#method.custom), to make those customizations and use the `Template` responder in routes. But, just like with `tera` in this instance, you should choose only one of the two approaches.
